README.md
# WinDoc python Interface für die Verarbeitung von Daten aus der Windoc Arztsoftware

Mit diesen Tools kann effizient auf einige Funktionen und Daten aus Windoc zugegriffen werden. Dieses Repositroy vereint häufig verwendete Funktionen und ist in mehreren anderen Repos eingebunden.


CHANGELOG.md

# Changelog

## v5.1

Fix for DOB parsing (other column)

## v5.0

Better code, more features.

Features with Intern:
* `address()`
* `zip()` code
* `city()`
* `svnr()`
* `sex()`

Features in db:
* `create_Intern`

Features in Kassenkartei:
* diagnose
* ecard
* impfung

## v4.3

Intern now has `height` and `weight` getter functions.

## v4.2

Intern now has `firstname`, `surname`, `sex` and `dob` getter functions.

## v4.1

LabTemplate does not require `Labsch` entry anymore.
`group` and `service` attributes may not exist.

## v4.0

**API Changes**
* LabTemplate now requires device identifier for 2nd argument
  Old: db context, analyte name
  New: db context, device identifier, analyte name
  Analyte names are for example: 'SO' for sofia, 'XN' for sysmex

## v3.1

Add method to get all phone numbers associated with Intern

## v3.0

Recode kassenkartei

* new class `klein_tools.kassenkartei.Kassenkartei` (represents Kartei for specific Intern)
* moved free function `klein_tools.kassenkartei.leistung` to new class
* added class to DB context manager
* added `klein_tools.kassenkartei.Kassenkartei.log` method
* added `klein_tools._db_impl.Intern.kartei` method to get Kassenkartei object for Intern

## v2.6

Fix DOB parsing

## v2.5

Added methods to Intern:
* fullname
* age

## v2.4

Workqueue update:
* Added unlock function to return task to queue

## v2.3

Added Queries.

## v2.2

Change telephone join type from JOIN to LEFT JOIN.
Patients are not required to have Stammzusatz record.

## v2.1

Fix missing import

## v2.0

Fix for Connection timeout problem.
Now requires all DB actions to be performed in with-context.

## v1.0

Initial release

Known issues:

* Connection is stored internally and times out.
  This causes the program to cease to function.


workqueue.py

"""This module handles the X_HOF_WORKQUEUE"""

#import logging
import pyodbc
import random

_db_handle = None
_lock_magic = None
#_log = logging.getLogger('queue')

def init(db: pyodbc.Connection):
    """Initializes the queue module

    Parameters:
        db - the pyodbc connection to use

    Returns: nothing

    Throws:
        pyodbc related exceptions will not be caught
    """
    global _db_handle
    global _lock_magic
    _db_handle = db

    c = db.cursor()
    found = 1
    while found > 0:
        _lock_magic = str(random.randint(0,9999))
        c.execute("SELECT Count(*) AS cnt FROM X_HOF_WORKQUEUE WHERE lock = ?", _lock_magic)
        found = c.fetchone().cnt
    c.close()

def peek(action: str, limit: int = 5):
    """Returns tasks from queue but does not mark them as locked

    Parameters:
        action - action key to look up items for
        limit - maximum count of items to return (default: 5)

    Returns: Array[pyodbc.Row]
        an array of queue elements
        Columns: id, origin, action, Intern, data

    Throws:
        pyodbc related exceptions will not be caught
    """
    assert type(limit) == int
    assert type(action) == str

    c = _db_handle.cursor()

    c.execute("SELECT id, origin, action, Intern, data FROM X_HOF_WORKQUEUE WHERE action = ? AND lock IS NULL AND error = 'F' LIMIT %d" % limit, action)

    result = c.fetchall()
    c.close()
    return result

def dequeue(action: str, limit: int = 5):
    """Fetches tasks from the queue

    Parameters:
        action - action key to look up items for
        limit - maximum count of items to return (default: 5)

    Returns: Array[pyodbc.Row]
        an array of queue elements
        Columns: id, origin, action, Intern, data

    Throws:
        pyodbc related exceptions will not be caught
    """
    assert type(limit) == int
    assert type(action) == str

    c = _db_handle.cursor()

    c.execute("SELECT id FROM X_HOF_WORKQUEUE WHERE action = ? AND lock IS NULL AND error = 'F' LIMIT %d" % limit, action)

    result = c.fetchall()
    if len(result) == 0:
        return []
    ids = ','.join([str(item.id) for item in result])

    c.execute("UPDATE X_HOF_WORKQUEUE SET lock = ? WHERE lock IS NULL AND id IN (%s)" % ids, _lock_magic)
    c.commit()

    c.execute("SELECT id, origin, action, Intern, data FROM X_HOF_WORKQUEUE WHERE id IN (%s) AND lock = ?" % ids, _lock_magic)
    result = c.fetchall()

    c.close()

    return result

def mark_error(d):
    """Marks a task as failed

    Parameters:
        d - either ID (int) or row (pyodbc.Row)

    Throws:
        pyodbc related exceptions will not be caught"""

    if type(d) == int:
        idnum = d
    elif type(d) == pyodbc.Row:
        idnum = d.id
    else:
        raise TypeError("type of parameter d is neither int nor pyodbc.Row")

    c = _db_handle.cursor()
    c.execute("UPDATE X_HOF_WORKQUEUE SET error = 'T' WHERE id = ?", idnum)
    c.commit()
    c.close()

def remove(task):
    """Removes a task from queue

    Parameters:
        d - either ID (int) or row (pyodbc.Row)

    Throws:
        pyodbc related exceptions will not be caught"""

    if type(task) == int:
        idnum = task
    elif type(task) == pyodbc.Row:
        idnum = task.id
    else:
        raise TypeError("type of parameter task is neither int nor pyodbc.Row")

    c = _db_handle.cursor()
    c.execute("DELETE FROM X_HOF_WORKQUEUE WHERE id = ? AND lock = ?", idnum, _lock_magic)
    c.commit()
    c.close()

def unlock(task_or_tasks):
    """Unlocks tasks making it available again for subsequent workers

    Parameters:
        task_or_tasks - either a task identifier (int|pyodbc.Row) or a list of them

    Throws:
        pyodbc related exceptions will not be caught"""

    if type(task_or_tasks) in [ int, pyodbc.Row ]:
        tasks = [ task_or_tasks ]
    elif type(task_or_tasks) == list:
        tasks = task_or_tasks
    else:
        raise TypeError("type of parameter task_or_tasks is neither a task identifier or a list")

    tasknums = []
    for t in tasks:
        if type(t) == int:
            tasknums.append(str(t))
        elif type(t) == pyodbc.Row:
            tasknums.append(str(t.id))
        else:
            raise TypeError("element in list parameter task_or_tasks has wrong type %s expecting int or pyodbc.Row" % repr(type(t)))

    c = _db_handle.cursor()
    c.execute("UPDATE X_HOF_WORKQUEUE SET lock = NULL WHERE id IN (%s) AND lock = ?" % ','.join(tasknums), _lock_magic)
    c.commit()
    c.close()


klein_tools.py

"""This module has tools to work with weird Klein-EDV data structures"""

import logging
_log = logging.getLogger('klein_tools')

import datetime

_db_handle = None

def init(db):
    global _db_handle
    _db_handle = db

class KTable:
    LUT = {
            '11': 'K10', # ÖGK-W
            '12': 'K10', # ÖGK-N
            '13': 'K10', # ÖGK-B
            '14': 'K10', # ÖGK-O
            '15': 'K10', # ÖGK-ST
            '16': 'K10', # ÖGK-K
            '17': 'K10', # ÖGK-S
            '18': 'K10', # ÖGK-T
            '19': 'K10', # ÖGK-V
            '05': 'K05', # BVAEB-EB
            '07': 'K07', # BVAEB-OEB
            '40': 'K40', # SVS-GW
            '50': 'K50', # SVS-LW
            '1A': 'K1A', # KFA Wien
            '4A': 'K4A', # KFA Linz
            '5A': 'K5A', # KFA Graz
            'PR': 'KPR', # Privat
            }

    def __init__(self, handle):
        assert type(handle) == str and len(handle) == 2
        self.num = handle
        self.table = KTable.LUT[handle]

    def position_from_service(self, serv):
        c = _db_handle.cursor()
        try:
            c.execute("SELECT * FROM %s WHERE Kurz = ?" % self.table, serv)
        except Exception as ex:
            print(self.num)
            print(repr(ex))
            raise
        res = c.fetchall()
        c.close()
        if not res or len(res) != 1:
            raise ValueError(f"Service {serv} returned erroneous count of results: {len(res)}")
        return res[0].Posnummer

class Intern:
    """Helper class to work with patients"""

    def insanify(i):
        """Returns the correctly padded Intern identifier

        Parameters:
            i - either int or str representation of Intern

        Returns:
            CHAR(6) Klein-style Intern identifier
        """
        iint = int(i)
        istr = str(iint)
        istr = ' ' * (6-len(istr)) + istr
        return istr

    def __init__(self, i):
        """Construct Intern object

        Parameters:
            i - CHAR(6) correctly padded Intern identifier"""
        self._data = {}
        self.Intern = i

    def exists(self):
        """Checks if this Intern exists in DB

        Returns:
            True if exists, False otherwise
        """
        c = _db_handle.cursor()
        c.execute("SELECT Intern FROM Stammdaten WHERE Intern = ?", self.Intern)
        res = c.fetchone()
        c.close()
        return res and len(res) == 1

    def phone(self):
        """Retrieve phone number

        Returns:
            some string representation of a phone number.
            Takes first match in columns with following order:
            Handy, Telefon, Tel1, Tel2"""
        if 'phone' in self._data:
            return self._data['phone']

        c = _db_handle.cursor()
        c.execute("SELECT Telefon, Tel1, Tel2, Handy FROM Stammdaten sd JOIN Stammzusatz sz ON sd.Intern = sz.Intern WHERE sd.Intern = ?", self.Intern)
        res = c.fetchone()
        c.close()

        if res is None:
            return None

        Telefon, Tel1, Tel2, Handy = res

        for attempt in [ (Handy, 'Handy'), (Telefon, 'Telefon'), (Tel1, 'Tel1'), (Tel2, 'Tel2') ]:
            if attempt[0] is not None and attempt[0].strip() != '':
                ret = attempt[0].strip()
                self._data['phone'] = ret
                return ret

        return None

    def kassen_ref(self):
        c = _db_handle.cursor()
        # encoding problems with "Versicherungsträger" should be column index 14
        c.execute("SELECT * FROM Stammdaten WHERE Intern = ?", self.Intern)
        res = c.fetchone()
        c.close()
        return KTable(res[14])

class Kassenkartei:
    def leistung(datum, pos, cnt, kasse, clock): #, price=None, clock=None, chef=None, pstat=None):
        if type(datum) == datetime.datetime:
            datum = datum.strftime("%Y%m%d")
        elif type(datum) == str:
            if len(datum) != 8:
                raise ValueError("argument datum expects CHAR(8) date value")
        elif type(datum) == int:
            datum = str(datum)
            if len(datum) != 8:
                raise ValueError("argument datum expects CHAR(8) date value")
        else:
            raise TypeError("argument datum has wrong type. Use one of datetime, str, int")

        if type(pos) != str:
            raise TypeError("argument pos must be a string")
        if len(pos) > 7:
            raise ValueError("argument pos exceeds maximum length of 7")
        pos += ' ' * (7 - len(pos))

        try:
            cnt = int(cnt)
        except TypeError:
            raise TypeError("argument cnt must be castable to integer")
        if cnt > 9999 or cnt < 0:
            raise ValueError("argument cnt needs to be at least 0 and less than 9999")
        cnt = str(cnt)
        #cnt = ' ' * (4 - len(cnt)) + cnt
        cnt += ' ' * (4 - len(cnt))

        if type(kasse) != str:
            raise TypeError("argument kasse must be a string")
        if len(kasse) != 2:
            raise ValueError("argument kasse must be CHAR(2) kassen ID")

        if not clock:
            clock = ' ' * 4
        if type(clock) == str:
            if len(clock) > 4:
                raise ValueError("argument clock must not exceed CHAR(4)")
            clock += ' ' * (4 - len(clock))
        elif type(clock) == datetime.datetime:
            clock = clock.strftime("%H%M")
        else:
            raise TypeError("argument clock must be a string or datetime object")

        pstat = ' ' * 8 # format ddmmYYYY
        price = ' ' * 10 # str repr of EUR
        chef1 = ' '
        chef2 = ' ' * 2 # CHEF CHAR(3)
        return f"{datum}{pos}    {clock}{cnt} {chef1}{pstat}{price}{chef2}{kasse}"

class LabTemplate:
    def __init__(self, sofia_order):
        self.lab = 'SO' + sofia_order
        # step 1: lookup in lab templates
        c = _db_handle.cursor()
        c.execute("SELECT * FROM Labsch WHERE Kurz = ?", self.lab)
        res = c.fetchall()
        c.close()
        if not res or len(res) != 1:
            raise ValueError("Lab reference '%s' has returned '%d' rows" % (self.lab, len(res)))
        self.service = res[0].KurzLK.strip()
        self.group = res[0].Grup
        if self.service == '':
            del self.service

def guess_if_positive(result):
    result = result.strip().lower()
    if result in ('negativ', 'neg', 'negative'):
        return False
    if result in ('positiv', 'pos', 'positive'):
        return True
    if 'pos' in result:
        return True
    if 'neg' in result:
        return False
    _log.error("result '%s' could not be guessed to be either positive or negative", result)
    return None

def ymd_pretty(ymd: str):
    return '.'.join([ymd[6:8], ymd[4:6], ymd[0:4]])



klein_tools/db.py

from . import _log as _parent_log
_log = _parent_log.getChild('db')

from . import _db_impl as _db
from . import kassenkartei as _kk

import pyodbc
import datetime

class StateError(Exception): pass

class Queries:
    def __init__(self, ctx):
        self._ctx = ctx

    def all_Interns_with_comms(self, with_condition=None):
        """Returns a list of preinitialized Intern objects

        Internally a SELECT query is executed with a single join.
        With the parameter with_condition a condition can be appended.
        Any query valid after a JOIN-ON cluase can be specified.
        Tables Stammdaten and Stammzusatz are already joined.

        Be aware of SQL-injection vulnerabilities.

        Parameters:
            with_condition - str, SQL query to append to filter results

        Returns:
            list[Intern]
        """
        c = self._ctx.unmanaged_cursor()
        query = "SELECT Stammdaten.Intern, Email, Telefon, Tel1, Tel2, Handy FROM Stammdaten LEFT JOIN Stammzusatz ON Stammdaten.Intern = Stammzusatz.Intern %s" % with_condition
        _log.getChild('Queries.all_Interns_with_comms').debug(query)
        c.execute(query)
        res = c.fetchall()
        c.close()

        ret = []
        for r in res:
            ii, Email, Telefon, Tel1, Tel2, Handy = r
            i = self._ctx.Intern(ii)
            i._find_phone(Telefon, Tel1, Tel2, Handy)
            if Email and Email.strip() != '':
                i._data['mail'] = Email
            else:
                i._data['mail'] = None
            ret.append(i)

        return ret

class Connection:
    def __init__(self, dsn):
        self._log = _log.getChild('Connection')
        self.dsn = dsn
        self._cleanup = []
        self.conn = None

    def __enter__(self):
        self.conn = pyodbc.connect(self.dsn)
        self._log.debug("Entered context, connected")
        return self

    def __exit__(self, exc_type, exc_val, exc_traceback):
        for c in self._cleanup:
            c.close()
        self._cleanup = []
        self.conn.close()
        self.conn = None
        self._log.debug("Left context, disconnected")

    def cursor(self):
        c = self.unmanaged_cursor()
        self._cleanup.append(c)
        return c

    def unmanaged_cursor(self):
        if self.conn:
            c = self.conn.cursor()
            return c
        else:
            raise StateError("Needs to be executed in with-context")

    def _managed(self, clazz, *args, **kwargs):
        if self.conn:
            obj = clazz(self, *args, **kwargs)
            self._cleanup.append(obj)
            return obj
        else:
            raise StateError("Needs to be executed in with-context")

    def Queries(self):
        return Queries(self)

    def KTable(self, handle):
        return _db.KTable(self, handle)

    def create_Intern(self, firstname, lastname, sex, dob, svnr, address, phone, kcode, county, category):
        SEX_LUT = {
                1 : '1',
                2 : '2',
                '1': '1',
                '2': '2',
                'M': '1',
                'W': '2',
                'm': '1',
                'w': '2'
                }
        c = self.unmanaged_cursor()
        if type(dob) == datetime.datetime:
            dob = dob.strftime('%d%m%Y')
        c.execute("EXECUTE createintern(?, ?, ?, ?, ?, ?, ?, ?, ?, ?)", firstname, lastname, SEX_LUT[sex], dob, svnr, address, phone, kcode, county, category)
        internid = c.fetchone()[0]
        c.commit()
        c.close()
        return _db.Intern(self, internid)

    def Intern(self, i):
        return _db.Intern(self , i)

    def LabTemplate(self, device, order):
        return _db.LabTemplate(self, device, order)

    def Kassenkartei(self, i):
        return _kk.Kassenkartei(i, ctx=self)

class Pool:
    def __init__(self, dsn):
        self.dsn = dsn

    def open(self):
        return Connection(self.dsn)


klein_tools/__init__.py

"""This module has tools to work with weird Klein-EDV data structures"""

import logging
_log = logging.getLogger('klein_tools')

def format_intern(i):
    """Returns the correctly padded Intern identifier

    Parameters:
        i - either int or str representation of Intern

    Returns:
        CHAR(6) Klein-style Intern identifier
    """
    iint = int(i)
    istr = str(iint)
    istr = ' ' * (6-len(istr)) + istr
    return istr

def guess_if_positive(result):
    result = result.strip().lower()
    if result in ('negativ', 'neg', 'negative'):
        return False
    if result in ('positiv', 'pos', 'positive'):
        return True
    if 'pos' in result:
        return True
    if 'neg' in result:
        return False
    _log.error("result '%s' could not be guessed to be either positive or negative", result)
    return None

def ymd_pretty(ymd: str):
    return '.'.join([ymd[6:8], ymd[4:6], ymd[0:4]])



klein_tools/kassenkartei.py
import datetime

from . import _log as _parent_log
_log = _parent_log.getChild('Kassenkartei')

class Kassenkartei:

    def __init__(self, Intern, ctx=None):
        self.Intern = Intern
        self._ctx = ctx

    @staticmethod
    def _handle_date(datum):
        if datum:
            if type(datum) == datetime.datetime:
                datum = datum.strftime("%Y%m%d")
            elif type(datum) == str:
                if len(datum) != 8:
                    raise ValueError("argument datum expects CHAR(8) date value")
            elif type(datum) == int:
                datum = str(datum)
                if len(datum) != 8:
                    raise ValueError("argument datum expects CHAR(8) date value")
            else:
                raise TypeError("argument datum has wrong type. Use one of datetime, str, int")
        else:
            datum = datetime.datetime.now().strftime('%Y%m%d')
        return datum

    @staticmethod
    def _handle_str(s, exactlen=None):
        if type(s) != str:
            raise TypeError("argument name must be a string not '%s'" % type(s))
        if exactlen:
            if len(s) > exactlen:
                raise ValueError(f"argument name may not exceed {exactlen} chars")
            s += ' ' * (exactlen - len(s))
        return s

    @staticmethod
    def make_diagnose(diag, kcode, datum=None):
        datum = Kassenkartei._handle_date(datum)
        diag = Kassenkartei._handle_str(diag, exactlen=78)
        eintr = f"{datum} {diag}{kcode}"
        return eintr

    def diagnose(self, diag, datum=None):
        c = self._ctx.unmanaged_cursor()
        datum = Kassenkartei._handle_date(datum)
        kcode = self._ctx.Intern(self.Intern).kassen_ref().num
        eintr = Kassenkartei.make_diagnose(diag, kcode, datum)
        _log.info(f"Intern={self.Intern} Date={datum} Kartei-Diagnose: {eintr}")
        c.execute("INSERT INTO Kassenkartei (Intern, Datum, Kennung, Eintragung) VALUES (?,?,?,?)", self.Intern, datum, 'D', eintr)
        c.commit()
        c.close()

    @staticmethod
    def make_ecard(idstr, withocard):
        prefix = 'o-card,' if withocard else ''
        return f"{prefix}{idstr}"

    def ecard(self, idstr, withocard, datum=None):
        c = self._ctx.unmanaged_cursor()
        datum = Kassenkartei._handle_date(datum)
        eintr = Kassenkartei.make_ecard(idstr, withocard)
        _log.info(f"Intern={self.Intern} Date={datum} Kartei-Ecard: {eintr}")
        c.execute("INSERT INTO Kassenkartei (Intern, Datum, Kennung, Eintragung) VALUES (?,?,?,?)", self.Intern, datum, 'E', eintr)
        c.commit()
        c.close()

    @staticmethod
    def make_impfung(name, charge, datum=None, erledigt=False):
        if type(name) != str:
            raise TypeError("argument name must be a string not '%s'" % type(name))
        if len(name) > 53:
            raise ValueError("argument name may not exceed 53 chars")
        charge = charge or ''
        datum = Kassenkartei._handle_date(datum)
        name += ' '*(53-len(name))
        erledigt = 'j' if erledigt else ' '
        return f"{datum}{name}{erledigt}{charge}"

    def impfung(self, name, charge, datum=None, erledigt=False):
        c = self._ctx.unmanaged_cursor()
        datum = Kassenkartei._handle_date(datum)
        eintr = Kassenkartei.make_impfung(name, charge, datum, erledigt)
        _log.info(f"Intern={self.Intern} Date={datum} Kartei-Impfung: {eintr}")
        c.execute("INSERT INTO Kassenkartei (Intern, Datum, Kennung, Eintragung) VALUES (?,?,?,?)", self.Intern, datum, 'I', eintr)
        c.commit()
        c.close()

    @staticmethod
    def make_log(msg):
        if type(msg) != str:
            raise TypeError("argument msg must be a string")
        if len(msg) > 80:
            raise ValueError("argument msg must not exceed 80 chars")

        return msg

    def log(self, msg, datum=None):
        c = self._ctx.unmanaged_cursor()
        datum = Kassenkartei._handle_date(datum)
        _log.info(f"Intern={self.Intern} Date={datum} Kartei-Log: {msg}")
        c.execute("INSERT INTO Kassenkartei (Intern, Datum, Kennung, Eintragung) VALUES (?,?,?,?)", self.Intern, datum, 'T', Kassenkartei.make_log(msg))
        c.commit()
        c.close()

    @staticmethod
    def make_leistung(datum, pos, cnt, kasse, clock): #, price=None, clock=None, chef=None, pstat=None):
        datum = Kassenkartei._handle_date(datum)

        if type(pos) != str:
            raise TypeError("argument pos must be a string")
        if len(pos) > 7:
            raise ValueError("argument pos exceeds maximum length of 7")
        pos += ' ' * (7 - len(pos))

        try:
            cnt = int(cnt)
        except TypeError:
            raise TypeError("argument cnt must be castable to integer")
        if cnt > 9999 or cnt < 0:
            raise ValueError("argument cnt needs to be at least 0 and less than 9999")
        cnt = str(cnt)
        #cnt = ' ' * (4 - len(cnt)) + cnt
        cnt += ' ' * (4 - len(cnt))

        if type(kasse) != str:
            raise TypeError("argument kasse must be a string")
        if len(kasse) != 2:
            raise ValueError("argument kasse must be CHAR(2) kassen ID")

        if not clock:
            clock = ' ' * 4
        if type(clock) == str:
            if len(clock) > 4:
                raise ValueError("argument clock must not exceed CHAR(4)")
            clock += ' ' * (4 - len(clock))
        elif type(clock) == datetime.datetime:
            clock = clock.strftime("%H%M")
        else:
            raise TypeError("argument clock must be a string or datetime object")

        pstat = ' ' * 8 # format ddmmYYYY
        price = ' ' * 10 # str repr of EUR
        chef1 = ' '
        chef2 = ' ' * 2 # CHEF CHAR(3)
        return f"{datum}{pos}    {clock}{cnt} {chef1}{pstat}{price}{chef2}{kasse}"

    def leistung(self, datum, pos, cnt, kasse, clock=None, override_ldatum=None):
        c = self._ctx.unmanaged_cursor()
        datum = Kassenkartei._handle_date(datum)
        eintr = Kassenkartei.make_leistung(override_ldatum or datum, pos, cnt, kasse, clock)
        _log.info(f"Intern={self.Intern} Entry-Date={datum} Sub-Date={override_ldatum or datum} Entry={eintr}")
        c.execute("INSERT INTO Kassenkartei (Intern, Datum, Kennung, Eintragung) VALUES (?,?,?,?)", self.Intern, datum, 'L', eintr)
        c.commit()
        c.close()



klein_tools/_db_impl.py

from .db import _log
from .kassenkartei import Kassenkartei

import datetime

class KTable:
    LUT = {
            '11': 'K10', # ÖGK-W
            '12': 'K10', # ÖGK-N
            '13': 'K10', # ÖGK-B
            '14': 'K10', # ÖGK-O
            '15': 'K10', # ÖGK-ST
            '16': 'K10', # ÖGK-K
            '17': 'K10', # ÖGK-S
            '18': 'K10', # ÖGK-T
            '19': 'K10', # ÖGK-V
            '05': 'K05', # BVAEB-EB
            '07': 'K07', # BVAEB-OEB
            '40': 'K40', # SVS-GW
            '50': 'K50', # SVS-LW
            '1A': 'K1A', # KFA Wien
            '4A': 'K4A', # KFA Linz
            '5A': 'K5A', # KFA Graz
            'PR': 'KPR', # Privat
            }

    def __init__(self, ctx, handle):
        assert type(handle) == str and len(handle) == 2
        self._ctx = ctx
        self.num = handle
        self.table = KTable.LUT[handle]

    def position_from_service(self, serv):
        c = self._ctx.unmanaged_cursor()
        try:
            c.execute("SELECT * FROM %s WHERE Kurz = ?" % self.table, serv)
        except Exception as ex:
            print(self.num)
            print(repr(ex))
            c.close()
            raise
        res = c.fetchall()
        c.close()
        if not res or len(res) != 1:
            raise ValueError(f"Service {serv} returned erroneous count of results: {len(res)}")
        return res[0].Posnummer

class Intern:
    """Helper class to work with patients"""

    def __init__(self, ctx, i):
        """Construct Intern object

        Parameters:
            i - CHAR(6) correctly padded Intern identifier"""
        self._ctx = ctx
        self._data = {}
        self.Intern = i

    def kartei(self):
        return Kassenkartei(self.Intern, ctx=self._ctx)

    def exists(self):
        """Checks if this Intern exists in DB

        Returns:
            True if exists, False otherwise
        """
        c = self._ctx.unmanaged_cursor()
        c.execute("SELECT Intern FROM Stammdaten WHERE Intern = ?", self.Intern)
        res = c.fetchone()
        c.close()
        return res and len(res) == 1

    def _find_phone(self, Telefon, Tel1, Tel2, Handy):
        self._data['phones'] = list(filter(None, [ x.strip() for x in [ Handy, Telefon, Tel1, Tel2 ] if x is not None ]))

        for attempt in [ (Handy, 'Handy'), (Telefon, 'Telefon'), (Tel1, 'Tel1'), (Tel2, 'Tel2') ]:
            if attempt[0] is not None and attempt[0].strip() != '':
                ret = attempt[0].strip()
                self._data['primary_phone'] = ret
                return ret

        self._data['primary_phone'] = None
        return None

    def _query_phone(self):
        c = self._ctx.unmanaged_cursor()
        c.execute("SELECT Telefon, Tel1, Tel2, Handy FROM Stammdaten sd LEFT JOIN Stammzusatz sz ON sd.Intern = sz.Intern WHERE sd.Intern = ?", self.Intern)
        res = c.fetchone()
        c.close()

        if res is None:
            Telefon, Tel1, Tel2, Handy = (None, None, None, None)
        else:
            Telefon, Tel1, Tel2, Handy = res

        return self._find_phone(Telefon, Tel1, Tel2, Handy)

    def phone(self):
        """Retrieve primary phone number

        Returns:
            some string representation of a phone number.
            Takes first match in columns with following order:
            Handy, Telefon, Tel1, Tel2"""

        if 'primary_phone' in self._data:
            return self._data['primary_phone']
        else:
            return self._query_phone()

    def phones(self):
        """Retrieve all associated phone numbers

        Returns:
            list of string representations of a phone numbers
        """

        if 'phones' not in self._data:
            self._query_phone()

        return self._data['phones']

    def _query_mail(self):
        c = self._ctx.unmanaged_cursor()
        c.execute("SELECT Email FROM Stammdaten WHERE Intern = ?", self.Intern)
        res = c.fetchone()
        c.close()

        mail = res.Email.strip()
        if mail == '':
            mail = None
        self._data['mail'] = mail
        return mail

    def mail(self):
        """Retrieve mail address

        Returns:
            some string representation of a mail address.
        """

        if 'mail' in self._data:
            return self._data['mail']
        else:
            return self._query_mail()

    def _query_data(self):
        c = self._ctx.unmanaged_cursor()
        c.execute("SELECT * FROM Stammdaten WHERE Intern = ?", self.Intern)
        res = c.fetchone()
        c.close()

        self._data['title'] = None
        title = res.Titel
        if title:
            title = title.strip()
            if title != '':
                self._data['title'] = title

        self._data['firstname'] = res.Vorname.strip()
        self._data['surname'] = res.Familienname.strip()

        self._data['address'] = res.Straße.strip()
        self._data['zip'] = res.Postleitzahl.strip()
        self._data['city'] = res.Wohnort.strip()

        self._data['dob'] = datetime.datetime.strptime(res.Beitragsnummer.strip(), "%d%m%Y")

        self._data['svnr'] = res.Versicherungsnummer.strip() + res.Geburtsdatum.strip()

        if res.Geschlecht == '1':
            self._data['sex'] = 'M'
        else:
            self._data['sex'] = 'F'

        self._data['height'] = res.Grösse.strip()
        self._data['weight'] = res.Gewicht.strip()

        for k in ['address', 'zip', 'city', 'height', 'weight']:
            if self._data[k] == '':
                self._data[k] = None

    def _get_data(self, key):
        if key not in self._data:
            self._query_data()
        return self._data[key]

    def address(self):
        return self._get_data('address')

    def zip(self):
        return self._get_data('zip')

    def city(self):
        return self._get_data('city')

    def svnr(self):
        return self._get_data('svnr')

    def sex(self):
        return self._get_data('sex')

    def firstname(self):
        return self._get_data('firstname')

    def surname(self):
        return self._get_data('surname')

    def fullname(self):
        if 'surname' not in self._data:
            self._query_data()

        hon = 'Herr' if self._data['sex'] == 'M' else 'Frau'
        title = self._data['title'] + ' ' if self._data['title'] else ''
        fir = self._data['firstname']
        sur = self._data['surname']

        return f"{hon} {title}{fir} {sur}"

    def height(self):
        return self._get_data('height')

    def weight(self):
        return self._get_data('weight')

    def dob(self):
        return self._get_data('dob')

    def age(self):
        now = datetime.datetime.now()
        diff = now - self.dob()
        
        return int(diff.days/365)

    def kassen_ref(self):
        c = self._ctx.unmanaged_cursor()
        # encoding problems with "Versicherungsträger" should be column index 14
        c.execute("SELECT * FROM Stammdaten WHERE Intern = ?", self.Intern)
        res = c.fetchone()
        c.close()
        return self._ctx.KTable(res[14])

class LabTemplate:
    def __init__(self, ctx, device, sofia_order):
        self._ctx = ctx
        self.lab = device + sofia_order
        # step 1: lookup in lab templates
        c = self._ctx.unmanaged_cursor()
        c.execute("SELECT * FROM Labsch WHERE Kurz = ?", self.lab)
        res = c.fetchall()
        c.close()
        if res:
            if len(res) != 1:
                raise ValueError("Lab reference '%s' has returned '%d' rows" % (self.lab, len(res)))
            service = res[0].KurzLK.strip()
            self.group = res[0].Grup
            if service != '':
                self.service = service

